Plugin/Controller/Cart/Add.php

```diff
@@ -0,0 +1,195 @@
+    ) {
+        $this->request = $request;
+        $this->customerSession = $session;
+        $this->collectionFactory = $collectionFactory;
+        $this->orderCollectionFactory = $orderCollectionFactory;
+        $this->checkoutSession = $checkoutSession;
+        $this->scopeConfig = $config;
+        $this->messageInterface = $messageInterface;
+        $this->resultRedirectFactory = $redirectFactory;
+    }
+
+    /**
+     * @param \Magento\Checkout\Controller\Cart\Add $subject
+     * @param callable $proceed
+     * @return mixed
+     */
+    public function aroundExecute(\Magento\Checkout\Controller\Cart\Add $subject, callable $proceed)
+    {
+        $usageId = $this->request->getParam('usage_id');
+        if ($this->request->getParam('usage_category') != $this->getCustomLicenseId()) {
+            return $proceed();
+        } elseif (array_key_exists($this->getCustomLicenseId(), $usageId)) {
+            if ($this->customerSession->isLoggedIn()) {
+                /**
+                 * @var \DevStone\UsageCalculator\Model\ResourceModel\MaxUsage\Collection $maxUsageColleciton
+                 */
+                $maxUsageColleciton = $this->collectionFactory->create()
+                    ->addFieldToSelect('max_usage')
+                    ->addFieldToFilter('usage_id', $usageId[$this->getCustomLicenseId()]);
+                if ($maxUsageColleciton->count()) {
+                    $maxUsage = $maxUsageColleciton->getFirstItem()['max_usage'];
+                    if ($maxUsage) {
+                        $totalUsageCountByOrder = $this->getUsageCountByOrders($usageId);
+                        if ($totalUsageCountByOrder < $maxUsage) {
+                            $totalUsageCountByQuote = $this->getUsageCountByQuote($usageId);
+                            if ($totalUsageCountByQuote < $maxUsage) {
+                                return $proceed();
+                            } else {
+                                $this->messageInterface->addErrorMessage(__('We can\'t add this item to your shopping cart right now Becuase you have reached you\'r  max usage limit'));
+                                return $this->resultRedirectFactory->create()->setPath('*/*/');
+
+                            }
+                        } else {
+                            $this->messageInterface->addErrorMessage(__('We can\'t add this item to your shopping cart right now Becuase you have reached you\'r  max usage limit'));
+                            return $this->resultRedirectFactory->create()->setPath('*/*/');
+
+                        }
+                    }
+                }
+            }
+        }
+        return $proceed();
+    }
+
+    /**
+     * @param $usageId
+     * @return int
+     */
+    public function getUsageCountByOrders($usageId)
+    {
+        $totalUsageCount = 0;
+        $ordersCollection = $this->orderCollectionFactory->create()
```

Comment:
This will have the same performance problems as mentioned above. Customers with many order may cause PHP to run out of memory. We need to find a way to do the counting in MySQL.

Reply:
@superdav42 I also had the same concern about memory timeout with this approach but if you check the structure of orders in database, items are stored in `sales_order_item` table. The information regarding usages are stored in `product_options` column which is JSON encoded so we can't check or load only those orders which have custom option `usage_category` to check `max_usages`.

Reply:
Let's just add a limit of 20.